201904-transpiler/charles/transpiler.example.js

```diff
@@ -0,0 +1,315 @@
+		}
+	}
+	if (code.match('export default')) {
+		return {
+			type: 'DefaultExport',
+		}
+	}
+	if (!isNaN(Number(code))) {
+		return {
+			type: 'Number',
+			value: code,
+		}
+	}
+	if (code.match('let')) {
+		return {
+			type: 'VariableDeclarator',
+		}
+	}
+	if (code.match(/[+-\/*]/gi)) {
+		return {
+			type: "BinaryOperator",
+			value: code,
+		}
+	}
+	if (code.match(/=/gi)) {
+		return {
+			type: 'VariableAssignmentOperator',
+		}
+	}
+	if (!code.includes('"')) {
+		return {
+			type: 'Identifier',
+			value: code,
+		}
+	}
+};
+
+/*
+	### PARSING ###
+*/
+
+const parser = tokens => {
+	/* ... */
+	return {
+		type: "Program",
+		statements: parseStatements(tokens),
+	}; // an object which is our AST - we can actually refer back to the tree traversal stuff we did
+};
+
+const performSyntacticAnalysis = tokens => {
+	const token = tokens[0];
+	const lookahead = tokens[1];
+	switch (tokens[0].type) {
+		case "DefaultExport": {
+			return parseDefaultExport(tokens);
+		}
+		case "VariableDeclarator": {
+			return parseVariableDeclaration(tokens);
+		}
+		case "Identifier": {
+			if (lookahead && lookahead.type === "VariableAssignmentOperator") {
+				return parseVariableAssigment(tokens);
+			}
+		}
+		case "String":
+		case "Number":  {
+			if (lookahead && lookahead.type === 'BinaryOperator') {
+				return parseBinaryExpression(tokens);
+			}
+			return tokens.shift();
+		}
+		default: {
+			tokens.shift();
+		}
+	}
+}
+
+const parseDefaultExport = tokens => {
+	tokens.shift();
+	return {
+		type: 'DefaultExportExpression',
+		value: performSyntacticAnalysis(tokens),
+	}
+}
+
+const parseVariableAssigment = tokens => {
+	let id = tokens.shift();
+	tokens.shift() //remove operator;
+	let value = performSyntacticAnalysis(tokens);
+
+	const expression = {
+		type: 'VariableAssignment',
+		id,
+		value,
+	}
+	console.log(expression);
+	return expression;
+}
+
+const parseStatements = tokens => {
+	const statements = [];
+	while (tokens.length > 0) {
+		statements.push(performSyntacticAnalysis(tokens));
+	}
+	return statements.filter(i=>i);
+}
+
+const parseBinaryExpression = tokens => {
+	let left = tokens.shift();
+	let operator = tokens.shift().value;
+	let right = performSyntacticAnalysis(tokens);
+
+	const expression = {
+		type: 'BinaryExpression',
+		left,
+		operator,
+		right,
+	}
+	return expression;
+}
+
+
+const parseValue = tokens => {
+	tokens.shift();
+	const value = tokens[0];
+	const lookahead = tokens[1]
+	if (tokens[0].value && !lookahead || lookahead.type === 'LineBreak' ) {
+		return tokens.shift();
+	} else if (lookahead.type === 'BinaryOperator') {
+		return parseBinaryExpression(tokens);
+	}
+}
+
+const parseVariableDeclaration = tokens => {
+	tokens.shift();
+	let id = tokens.shift();
+	let value = parseValue(tokens);
+	return {
+		type: "VariableDeclaration",
+		id,
+		value,
+	}
+}
+
+/*
+	## TRANSFORMATION ##
+*/
+const constructWithReferences = AST => AST;
+const transformer = AST => {
+	/* ... */
```

Comment:
@Noviny time complexity here is pretty terrible, since we're doing two traversals. Not really sure of a good way around this, suggestions welcome